gax/src/main/java/com/google/api/gax/rpc/QueuingResponseObserver.java

```diff
@@ -57,29 +57,29 @@
 
   private final BlockingQueue<Object> buffer = Queues.newArrayBlockingQueue(2);
   private StreamController controller;
-  private boolean isCancelled;
+  private volatile boolean isDone;
 
   void request() {
     controller.request(1);
   }
 
   Object getNext() throws InterruptedException {
-    if (isCancelled) {
+    if (isDone) {
```

Comment:
This is entirely my fault. I think you need to go back to your original impl where cancel and isDone are separate. They have different semantics:
- a cancelled stream won't return anymore elements even if the buffer is non-empty
- a done stream will still return elements after onComplete as long as there are elements in the buffer.

Can you:
- rename isDone to upstreamExhausted
- bring back isCancelled
- add a comment documenting the difference between the two
- add a test that ensures that a stream that has elements in the buffer will still return them when onComplete is called

Reply:
> add a test that ensures that a stream that has elements in the buffer will still return them when onComplete is called

I think this is tested in testMultipleItemStream https://github.com/googleapis/gax-java/blob/main/gax/src/test/java/com/google/api/gax/rpc/ServerStreamTest.java#L78